### PR TITLE
Add LLM pretraining tutorial

### PR DESCRIPTION
**Pull request by windmaple**

### CODE REVIEW

File: docs/JAX_for_LLM_pretraining.md [hunk 1]
```diff
@@ -0,0 +1,448 @@
+---
+jupyter:
+  jupytext:
+    text_representation:
+      extension: .md
```
Comment (jakevdp) [inline docs/JAX_for_LLM_pretraining.md:5]: Can you run the commands at https://jax-ai-stack.readthedocs.io/en/latest/contributing.html#adding-a-new-notebook to use `md:myst` format for the markdown?

File: docs/JAX_for_LLM_pretraining.md [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+
+To get a detailed understanding of how JAX automatic parallelism works, please refer to this [JAX tutorial](https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html#way-batch-data-parallelism-and-2-way-model-tensor-parallelism). In our case to leverage parallelism, we first need to define a `Mesh`, which declares the TPU resources with 2 axes: `batch` axis as 4 and `model` axis as 2, which maps to the TPU v2 cores. Here, the `model` axis enables the tensor parallel for us.
```
Comment (jakevdp) [inline docs/JAX_for_LLM_pretraining.md:87]: Let's link to the intro parallelism doc instead: https://jax.readthedocs.io/en/latest/sharded-computation.html It's a somewhat gentler introduction.

File: docs/JAX_for_LLM_pretraining.md [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    metrics.update(loss=loss, logits=logits, lables=batch[1])
+    optimizer.update(grads)
+```
+
+<!-- #region id="5um2vkeUNckm" -->
+Start training. It takes ~50 minutes on Colab.
+
+Note that for data parallel, we are sharding the training data along the `batch` axis using `jax.device_put` with `NamedeSharding`.
+
+We are also using the `jax.vamp` transformation to produce the target sequences faster.
```
Comment (jakevdp) [inline docs/JAX_for_LLM_pretraining.md:335]: Typo: `jax.vmap`

**Review by jakevdp — CHANGES_REQUESTED** [review]
Great content overall! A few minor comments and then I'd love to get this merged 😁

File: docs/JAX_for_LLM_pretraining.md (hunk 2 above)
Comment (windmaple) [reply docs/JAX_for_LLM_pretraining.md]: done

File: docs/JAX_for_LLM_pretraining.md (hunk 3 above)
Comment (windmaple) [reply docs/JAX_for_LLM_pretraining.md]: fixed

File: docs/JAX_for_LLM_pretraining.md (hunk 1 above)
Comment (windmaple) [reply docs/JAX_for_LLM_pretraining.md]: I had some issues w/ ruff, but hopefully I did it right sync'ing md/notebook.

File: docs/JAX_for_LLM_pretraining.md [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+id: 6zMsOIc7ouCO
+outputId: b16b0781-d1bd-456a-9757-533562f581df
+---
+!pip install jax-ai-stack
+!pip install -U tiktoken grain matplotlib
```
Comment (jakevdp) [inline docs/JAX_for_LLM_pretraining.md:54]: This `pip install` generates a lot of log text that is not particularly relevant to readers of the tutorial. You could add `tags: [hide-output]` in the metadata section for this cell to hide that in the rendered version.

If you do this, you may wish to split the `jax.devices()` call into another cell

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: docs/JAX_for_LLM_pretraining.md [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3
+  name: python3
+---
+
++++ {"id": "NIOXoY1xgiww"}
+
+### Copyright 2024 Google LLC.
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply docs/JAX_for_LLM_pretraining.md:16]: Also, the reason the doc build is failing is because the heading here is starting with H3.

File: docs/JAX_for_LLM_pretraining.md (hunk 4 above)
Comment (windmaple) [reply docs/JAX_for_LLM_pretraining.md]: I added -q to pip install. Also split the jax.devices into a new cell

File: docs/JAX_for_LLM_pretraining.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (windmaple) [reply docs/JAX_for_LLM_pretraining.md]: Added a H1 title

File: docs/JAX_for_LLM_pretraining.md [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+
+Remember in cell #5, we use 4-way data parallel and 2-way tensor parallel. Of course there are different ways to partition your model/data. For example, 8-way data parallel is another popular way. To switch to 8-way data parallel, uncomment the last line in cell # 4 to replace the `Mesh` definition with:
+
+`mesh = Mesh(mesh_utils.create_device_mesh((8, 1)), ('batch', 'model'))`
+
+JAX will automatically figure out how to shard the model and data to use the new partition strategy and nothing else need to be done. Re-connect the TPU runtime and run it again to see how it runs.
+
+How simple and powerful is this! And that's the beauty of JAX automatic parallelism.
+
+```{code-cell}
```
Comment (jakevdp) [inline docs/JAX_for_LLM_pretraining.md:487]: Let's delete the empty code cell

**Review by jakevdp — APPROVED** [review]
Looks great! One last comment below.

After addressing that, could you please squash the changes into a single commit, and then we'll get this merged!

**Comment by windmaple** [comment]
Might be easier if you choose 'Squash and merge' in the Merge dropdown menu?

> Looks great! One last comment below.
> 
> After addressing that, could you please squash the changes into a single commit, and then we'll get this merged!

**Comment by jakevdp** [comment]
> Might be easier if you choose 'Squash and merge' in the Merge dropdown menu?

Yeah I'll do that – sorry that response was force of habit: in the JAX repo it's not possible for the reviewer to squash commits.

**Comment by jakevdp** [comment]
Please fix the lint errors – you can do this locally by running `pre-commit run jupytext --all-files`

**Comment by windmaple** [comment]
lint errors fixed

File: docs/JAX_for_LLM_pretraining.md [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+plt.show()
+```
+
++++ {"id": "WB-ExEt1Zl1C"}
+
+As you can see, the model goes from generating completely random words at the beginning to generating sensible tiny stories at the end of the training. So essentially we have pretrained a small LLM to write tiny stories for us.
+
++++ {"id": "soPqiR1JNmjf"}
+
+# Saving
```
Comment (jakevdp) [inline docs/JAX_for_LLM_pretraining.md:444]: Here and below, we shouldn't use top-level headings except at the top level. Sorry for the late review here: I noticed this when looking at the [rendered table of contents](https://jax-ai-stack--81.org.readthedocs.build/en/81/tutorials.html) once all the build errors were fixed.

**Review by jakevdp — APPROVED** [review]
Awesome - let's get it in!
